src/common/cvt.cpp

```diff
@@ -1033,6 +1058,964 @@ void CVT_string_to_datetime(const dsc* desc,
+				}
+
+				const int fractionsPrecision = fractions != 0 ? std::log10(fractions) + 1 : 1;
+				int additionalZerosCount = 0;
+				if (number > fractionsPrecision)
+				{
+					additionalZerosCount = number - fractionsPrecision;
+					number = fractionsPrecision;
+				}
+
+				patternResult.printf("%d%.*s", fractions / (int) powf(10, fractionsPrecision - number),
+					additionalZerosCount, "00000000");
+			}
+			break;
+
+		case 'T':
+			date_type_check(ExpectedDateType::TIMEZONE, desc, pattern, cb);
+
+			if (pattern == "TZH")
+			{
+				SSHORT timezoneOffset = extract_timezone_offset(desc);
+				int timezoneSign = sign(timezoneOffset);
+				SSHORT offsetInHours = abs(timezoneOffset / 60);
+
+				string printfFormat = "%02d";
+
+				if (previousPattern != "TZM")
+				{
+					if (timezoneSign < 0)
+						printfFormat = "-" + printfFormat;
+					else
+						printfFormat = "+" + printfFormat;
+				}
+
+				patternResult.printf(printfFormat.c_str(), offsetInHours);
+			}
+			else if (pattern == "TZM")
+			{
+				SSHORT timezoneOffset = extract_timezone_offset(desc);
+				int timezoneSign = sign(timezoneOffset);
+				SSHORT offsetInMinutes = abs(timezoneOffset % 60);
+
+				string printfFormat = "%02d";
+
+				if (previousPattern != "TZH")
+				{
+					if (timezoneSign < 0)
+						printfFormat = "-" + printfFormat;
+					else if (timezoneOffset >= 0)
+						printfFormat = "+" + printfFormat;
+				}
+
+				patternResult.printf(printfFormat.c_str(), offsetInMinutes);
+			}
+			else if (pattern == "TZR")
+				patternResult = extract_timezone_name(desc);
+			break;
+
+		default:
+			invalid_pattern_exception(pattern, cb);
+	}
+
+	if (patternResult.isEmpty())
+		invalid_pattern_exception(pattern, cb);
+
+	return patternResult;
+}
+
+
+string CVT_datetime_to_format_string(const dsc* desc, const string& format, Callbacks* cb)
+{
+	if (format.isEmpty())
+		cb->err(Arg::Gds(isc_sysf_invalid_null_empty) << Arg::Str(STRINGIZE(format)));
+
+	struct tm times;
+	memset(&times, 0, sizeof(struct tm));
+
+	int fractions = 0;
+
+	switch (desc->dsc_dtype)
+	{
+		case dtype_sql_time:
+			Firebird::TimeStamp::decode_time(*(GDS_TIME*) desc->dsc_address,
+				&times.tm_hour, &times.tm_min, &times.tm_sec, &fractions);
+			break;
+
+		case dtype_sql_time_tz:
+		case dtype_ex_time_tz:
+			TimeZoneUtil::decodeTime(*(ISC_TIME_TZ*) desc->dsc_address,
+				true, TimeZoneUtil::NO_OFFSET, &times, &fractions);
+			break;
+
+		case dtype_sql_date:
+			Firebird::TimeStamp::decode_date(*(GDS_DATE*) desc->dsc_address, &times);
+			break;
+
+		case dtype_timestamp:
+			Firebird::TimeStamp::decode_timestamp(*(GDS_TIMESTAMP*) desc->dsc_address, &times, &fractions);
+			break;
+
+		case dtype_timestamp_tz:
+		case dtype_ex_timestamp_tz:
+			TimeZoneUtil::decodeTimeStamp(*(ISC_TIMESTAMP_TZ*) desc->dsc_address,
+				true, TimeZoneUtil::NO_OFFSET, &times, &fractions);
+			break;
+
+		default:
+			cb->err(Arg::Gds(isc_invalid_data_type_for_date_format));
+	}
+
+	string formatUpper(format);
```

Comment:
```suggestion
	const string formatUpper(format);
```
As you save it in `string_view`.

Reply:
Sorry, I see you change it. Hope it's never reallocated.

Reply:
It shouldn't